### PR TITLE
CR-1113174 : Added support for xbmgmt2 hotplug command

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
CR-1113174 : Need support for xbmgmt hotplug command with latest XRT

Issue:
In near future we are going to deprecate the xbmgmt command. Hence, we may need to add this support to the xbmgmt2.

Solution:
Added the required changes in xbmgmt2 tools.
Hotplug offline, Implemented as a query as it is Linux specific. Hotplug online, Implemented locally. As online doesn't have any device attached.

Test:
We have done the hotplug testing with U.2 card.

New Command structure:
Hotplug offline:
xbmgmt2 advanced --hotplug --action offline --device 0000:62:00.0

hotplug online:
xbmgmt2 advanced --hotplug --action online

### CODE REVIEW

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 1]
```diff
@@ -350,6 +350,22 @@ struct instance
...
+  using result_type = query::hotplug_offline::result_type;
+
+  static result_type
+  get(const xrt_core::device* device, key_type)
+  {
+    auto mgmt_dev = pcidev::get_dev(device->get_device_id(), false);
+
+    // Remove both user_pf and mgmt_pf
+    if (pcidev::shutdown(mgmt_dev, true, true))
+      throw xrt_core::query::sysfs_error("Hotplug offline Failed");
```
Comment (AShivangi) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:364]:
```suggestion
      throw xrt_core::query::sysfs_error("Hotplug offline failed");
```

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+static void
+hotplug_online()
+{
+  static const std::string  rescan_path = "/sys/bus/pci/rescan";
+
+  if (!boost::filesystem::exists(rescan_path))
+    throw xrt_core::error((boost::format("Invalid sysfs file path '%s'.") % rescan_path).str());
+
+  std::ofstream rescan_file(rescan_path, std::ofstream::out);
+  try
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp:48]: Why is there a try here if you are re-throwing the error? I don't think it's needed here

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    if (!boost::iequals(m_action, "online") && !boost::iequals(m_action, "offline")) {
+      std::cerr << boost::format("ERROR: Invalid action value: '%s'\n") % m_action;
+      printHelp();
+      throw xrt_core::error(std::errc::operation_canceled);
+    }
+
+    const bool is_online = boost::iequals(m_action, "online");
+
+    // Device BDF need to spcify for offline (hot removal) case
+    if (!is_online) {
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp:135]: Just a personal suggestion. This makes code easier to read :) 
```suggestion
    const bool is_offline = boost::iequals(m_action, "online");

    // Device BDF need to spcify for offline (hot removal) case
    if (is_offline) {
```
Or switch the if/else statements

**Review by AShivangi — CHANGES_REQUESTED** [review]
Looks pretty good! Minor comments :)

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+#include "OO_Hotplug.h"
+#include "tools/common/XBUtilitiesCore.h"
+#include "tools/common/XBUtilities.h"
+
+// XRT - Include Files
+#include "core/common/query_requests.h"
+#include "core/common/system.h"
+
+// 3rd Party Library - Include Files
+#include <boost/filesystem.hpp>
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp:29]: List should be shorted.

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
+  if (m_help) {
+    printHelp();
+    return;
+  }
+
+  // Exit if neither action or device specified
+  if (m_action.empty()) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp:119]: Hmm...If action is required to be spectified with a value, does that make this conditional statement obsolete for boost::program_options is peforming this check.

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp]: Resolved. Thanks

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp]: Updated. Thanks

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp (hunk 5 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp]: Removed. Thanks

File: src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp:29]: Each category is sorted individually.

**Review by chvamshi-xilinx — APPROVED** [review]

**Review by rbramand-xilinx — APPROVED** [review]

**Review by AShivangi — APPROVED** [review]

**Review by stsoe — APPROVED** [review]

**Review by rozumx — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by saifuddin-xilinx** [comment]
Retest this please. Fail is not related to my changes. Infra issue I believe.

**Pull request merged**